evm/src/NttManager/NttManagerWithPerChainTransceivers.sol

```diff
@@ -27,61 +73,184 @@ contract NttManagerWithPerChainTransceivers is NttManagerNoRateLimiting {
-        _bitmaps[forChainId].bitmap |= uint64(1 << transceiverInfo.index);
+            _getPerChainTransceiverBitmapStorage(RECV_TRANSCEIVER_BITMAP_SLOT);
+        uint64 oldBitmap = _bitmaps[forChainId].bitmap;
+        _bitmaps[forChainId].bitmap = indexBitmap;
+
+        // Update the thresold.
+        mapping(uint16 => _Threshold) storage _threshold = _getThresholdStoragePerChain();
+        uint8 oldThreshold = _threshold[forChainId].num;
+        _threshold[forChainId].num = threshold;
+
+        // Update the chain list.
+        if (indexBitmap == 0) {
+            _removeChain(RECV_ENABLED_CHAINS_SLOT, forChainId);
+        } else {
+            _addChainIfNeeded(RECV_ENABLED_CHAINS_SLOT, forChainId);
+        }
+
+        emit RecvTransceiversUpdatedForChain(
+            forChainId, oldBitmap, indexBitmap, oldThreshold, threshold
+        );
+    }
+
+    /// @notice Sets the transceiver bitmaps and thresholds for a set of chains.
+    /// @param params The values to be applied for a set of chains.
+    function setTransceiversForChains(
+        SetTransceiversForChainEntry[] memory params
+    ) external onlyOwner {
+        uint256 len = params.length;
+        for (uint256 idx = 0; idx < len;) {
+            setSendTransceiverBitmapForChain(params[idx].chainId, params[idx].sendBitmap);
+            setRecvTransceiverBitmapForChain(
+                params[idx].chainId, params[idx].recvBitmap, params[idx].recvThreshold
+            );
+            unchecked {
+                ++idx;
+            }
+        }
     }
 
+    // =============== Internal Interface Overrides ===================================================
+
     function _isSendTransceiverEnabledForChain(
         address transceiver,
         uint16 forChainId
     ) internal view override returns (bool) {
         uint64 bitmap = _getPerChainTransceiverBitmap(forChainId, SEND_TRANSCEIVER_BITMAP_SLOT);
         uint8 index = _getTransceiverInfosStorage()[transceiver].index;
-        return (bitmap & uint64(1 << index)) != 0;
+        return (bitmap & uint64(1 << index)) > 0;
+    }
+
+    /// @inheritdoc IManagerBase
+    function isMessageApproved(
+        bytes32 digest
+    ) public view override(ManagerBase, IManagerBase) returns (bool) {
+        uint16 sourceChainId = _getMessageAttestationsStorage()[digest].sourceChainId;
+        uint8 threshold = getThresholdForChain(sourceChainId);
+        return messageAttestations(digest) >= threshold && threshold > 0;
+    }
+
+    function _getMessageAttestations(
+        bytes32 digest
+    ) internal view override returns (uint64) {
+        AttestationInfo memory attInfo = _getMessageAttestationsStorage()[digest];
+        uint64 enabledTransceiverBitmap = _getEnabledTransceiversBitmap();
+        uint64 enabledTransceiversForChain =
+            _getEnabledRecvTransceiversForChain(attInfo.sourceChainId);
+        return attInfo.attestedTransceivers & enabledTransceiverBitmap & enabledTransceiversForChain;
```

Comment:
The `_checkTransceiversInvariants()` ensures that you can't remove a transceiver globally, when it's enabled for a chain. Therefore `enabledTransceiversForChain` is a subset of `enabledTransceiverBitmap`.